docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,109 @@
+# Freeing objects and resources
+
+* Status: Propsed
+* Deciders: ?
+* Date: 2023-10-10
+
+Prevous discussion: [PR 1394](https://github.com/mozilla/uniffi-rs/issues/1394), [Issue 8](https://github.com/mozilla/uniffi-rs/issues/8)
+
+## Context and Problem Statement
+
+UniFFI must support objects that hold resources like database connections, file handles, network sockets, etc.  This is difficult, because objects like this are handled differently by different languages:
+
+* On Kotlin and other GC-based languages, the general advice is to explicitly close the resources and not rely on garbage collection to deal with them.
+* On Rust and other RAII-based languages, the general advice is to close the resources in a destructor.
+
+Currently, our Kotlin bindings try to address this problem for Objects by adding a `destroy()` method that
+manually drops the Rust value and implementing the `AutoCloseable` interface which allows users to
+ensure this is called via the `use` extension function.  However, this is not sufficient for several
+reasons:
+
+  * It is specific to Kotlin, but many languages face similar issues.
+  * All objects get this handling, when only some need it.
+  * This system is tied to decision to *not* support freeing UniFFI objects via GC-based systems
+    like `Cleaner`, which is not correct for some interfaces (see #1394)
```

Comment:
I linked #1394 for the opposite reason: to show that sometimes we should rely on the GC.  The sentence is confusing because there's a weird double-negative in there though.  I'll try to clean that up.